### DISCUSSION TITLE
Time series analysis

### DISCUSSION THREAD

**Discussion by yungyuc** [solvcon/modmesh#380] [Ideas]
@j8xixo12 brought up an idea of implementing time-series analysis (in frequency domain) in modmesh. It consists of two parts: time-series and integral transform (analysis).

It is interesting to develop both from ground up. Existing (and mature) third-party packages can be used as references during the development.

@j8xixo12 do you have any ideas of how we may start the work?

**Comment by j8xixo12** [comment]
In my opinion, there are three parts we can start to work:
- File loader
    - Support csv file <-- Most simple and common format.
- Analysis algorithm implementation
    - FFT implementation
- Integrate with GUI
    - Select which column needs to be plotted in control panel
    - Select and apply filter or some analysis techniques in a configuration window <-- I'm not sure how to do it.

**Reply by yungyuc** [reply]
What would be the amount / size of data we may shoot for in the beginning?

**Reply by j8xixo12** [reply]
I think we can start with processing hundreds of thousand of data.

**Reply by yungyuc** [reply]
Anything less than a million entries is tiny.  It will be a good model problem.

**Reply by j8xixo12** [reply]
I'm not sure what is model problem ?
Do you mean FFT is a good problem to demonstrate array library of `modmesh` ?

**Reply by yungyuc** [reply]
Yes, FFT and any use of less than a million entries is a good problem to develop the code.  A model problem is a problem suitable for code development.

**Reply by j8xixo12** [reply]
Got it, I think I can try to implement the FFT with modmesh arrary library.
I also need to study how to write test code to ensure the implementation of FFT is correct.

**Reply by yungyuc** [reply]
We can use existing packages (including) numpy to verify.

**Reply by tigercosmos** [reply]
First we need to implement the complex system. I am now studying from Numpy

**Reply by yungyuc** [reply]
A complex-numbered array is good to have.  Without it we can simply use two arrays for the real and imaginary parts to start coding.

**Comment by j8xixo12** [comment]
If we need real data for validation, here is the [NASA official data portal](https://data.nasa.gov/) that allows everybody to download some real data.

**Reply by yungyuc** [reply]
@j8xixo12 Would https://techport.nasa.gov/view/116144 be useful for prototyping time-series code?

**Reply by j8xixo12** [reply]
Yes, these real flight data would be useful for prototyping time-series code!
We can use these real data which contains multiple frequencies and noise to test our time-series code.

**Reply by yungyuc** [reply]
Kudos to @xingularity for adding the first prototype with PR #431.

@xingularity , after PR #431, can you use the prototype to solve the problem described in https://techport.nasa.gov/projects/116144 ?  If not, what still miss?

**Comment by xingularity** [comment]
Hi @yungyuc  and @j8xixo12 

I think this idea containing two parts, one is to build a container (or we can call it time series DataFrame) to preserve and process time series data. The other one is to perform Fourier transform. Regarding the container part, I had a short discussion with you on Saturday. I think I can summarize below.

- Create a Time Series DataFrame to store time series data. This data frame will have an index column which stores the timestamp/epoch, and multiple columns to store time series data corresponding to these timestamps.
- This DataFrame should have capabilities to append one row or multiple rows at a time. The usage scenario of this feature would be like sensors getting one or multiple new records and these new data were appended onto this data frame (memory management consideration needed).
- Analysis capabilities. For the most simple one, user can extract one arbitrary column as a single time series, and then this time series can be passed into some Fourier transform (or other time series analysis) functions to perform analysis. 
  * [Extended Idea] Do we need another container for time series data structure? It should contain one time series data and an index column (time stamp). If the extracted series is only for Fourier analysis,  then we don't need it, just using SimpleArray is sufficient. Since transformed series is in frequency domain and there will be no time stamp.
- [Extended Idea] Does this Time Series DataFrame need to have abilities to attach one more series as a new column?  
  * If this DataFrame can attach new columns, then we need to consider merging two time series. If we need to consider attach other time series onto an existing DataFrame. Then we need to consider merging two different index columns into one (The attaching series may not have exact same time stamps as the original data frame), and deal with null values in the DataFrame (there is null data in a column on certain time stamp).

**Reply by yungyuc** [reply]
It is too quick to delve into the implementation plan.  I agree we will need a container, but it is obvious that processing data needs a container.

It is good to dissect the requirements into the bullets.  But the very first thing to do is to make a prototype for the real model problem listed in https://github.com/solvcon/modmesh/discussions/380#discussioncomment-9979376 .  It will give us a concrete sense of what should be done in the first place.

@xingularity would you please make an estimate for when you can work out some code processing the NASA dataset?

**Reply by xingularity** [reply]
I think I can make a prototype by this weekend to have a container which can 
- Read and store one CSV into a DataFrame with an index column (to be epoch or long int).
- Store multiple columns.

**Reply by yungyuc** [reply]
Perhaps the "data frame" may be prototyped by using a Python object holding [`SimpleArray`](https://github.com/solvcon/modmesh/blob/180e91122977d6a75d3c0cc6935cd744f0edf22a/cpp/modmesh/buffer/pymod/wrap_SimpleArrayPlex.cpp#L191).

**Reply by xingularity** [reply]
@yungyuc  

I think we can do this as long as we haven't started to worry about the memory management issue.

**Reply by yungyuc** [reply]
Right.  In a prototype we do not emphasize performance (runtime and memory).  Profiling will be needed after the prototype proves the correctness.

**Reply by xingularity** [reply]
> It is too quick to delve into the implementation plan. I agree we will need a container, but it is obvious that processing data needs a container.
> 
> It is good to dissect the requirements into the bullets. But the very first thing to do is to make a prototype for the real model problem listed in [#380 (reply in thread)](https://github.com/solvcon/modmesh/discussions/380#discussioncomment-9979376) . It will give us a concrete sense of what should be done in the first place.
> 
> @xingularity would you please make an estimate for when you can work out some code processing the NASA dataset?

I am not trying to quickly dive into the implementation plan. I just tried to record down everything we have discussed and thoughts to avoid losing them. We can do a prototype first and improve them gradually. Good reminder though, thanks!

**Reply by yungyuc** [reply]
We are aligned.

**Reply by j8xixo12** [reply]
It seems like merging time series or append column data into DataFrame could be useful in financial field. However implementation this feature might more complex. I think we can lower its priority during the prototyping.

**Reply by yungyuc** [reply]
We should start small and grow.  A key point for initial development is to quickly understand the problem.  Python shines for this kind of prototyping.  The architecture of modmesh has been curated to be productive to take Python prototypes into C++ and connect them together, and enables very fast learning through doing.

**Comment by xingularity** [comment]
Hi @yungyuc , @tigercosmos  and @j8xixo12 

Thanks yyc for merging my early prototype. However, the work has not done yet. Based on my f2f discussion with Chun-Hsu, he needs that this DataFrame to provide data in correct sequence. Current prototype implementation only reads text and organizes data in a columnar format, it does not guarantee the sequence of the data. So the next step should be implementing a sorting algorithm to check/guarantee the data sequence. There are two ways to put in the sorting algorithm I can think of, one is to sort data when the data are read or inserted, another one is to sort data only when they are retrieved. Any thoughts on this?

Collected backlog so far
- use mmap when reading large files. - yyc
- Index sequence should accept floating points. - tigercosmos
- The code will be more readable to have member data listed in __init__(). - yyc

**Reply by yungyuc** [reply]
Thanks, @xingularity .  Could you please create an issue to track the data ordering requirement?  It may need clarification.

`mmap` is a distant item.  We do not need to consider it before the data frame code is moved to C++.

**Comment by xingularity** [comment]
Hi @yungyuc  and @j8xixo12 
During my search, I found this document describing NASA standard of trending analysis. It should be helpful for our development. I record the link here.
https://ntrs.nasa.gov/citations/19910013502
